hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java

```diff
@@ -3069,27 +3073,33 @@ private RegionScannerImpl getScanner(Scan scan, List<KeyValueScanner> additional
           checkFamily(family);
         }
       }
-      return instantiateRegionScanner(scan, additionalScanners, nonceGroup, nonce);
+      return instantiateRegionScanner(scan, additionalScanners, additionalScannersForStores,
```

Comment:
Whats the diff between the two additional Scanner?

Reply:
Let me explain this.

There was one requirement where the additional cells of a mutation need to be visible for followed mutations. For example, when we have a Put operation and a CheckAndMutate operation in a single mutation batch, the additional cells of the Put operation should be visible for the followed CheckAndMutate operation.

The following code keeps additional cells of mutations:
https://github.com/brfrn169/hbase/blob/HBASE-24996/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java#L3833-L3835

To achieve this, I first tried to use the existing additionalScanners of RegionScannerImpl:
https://github.com/brfrn169/hbase/blob/HBASE-24996/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java#L3061

However, I found it's much easier to add a new additionalScanners to StoreScanner and use it than to use the existing additionalScanners of RegionScannerImpl.

So, I decided to add additionalScanners to StoreScanner and was able to make the additional cell visible for the followed mutations.
https://github.com/brfrn169/hbase/blob/HBASE-24996/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreScanner.java#L251-L253

By the way, the test code of this is as as follows:
https://github.com/brfrn169/hbase/blob/HBASE-24996/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestHRegion.java#L7101-L7174

Thanks.

Reply:
How do process increment and append in the past? What if we have a put and then an increment on the same row before this PR?

Reply:
We didn't handle this situation in the past. So if we have a put and then an increment on the same row, it acts like the put is ignored.

Reply:
So let's not mix things up in a single PR? Open another PR to discuss this first? I do not think this is the only way to do this. For HBase, it is not like the traditional RDBMS, where we issue a start transaction command to the server first, and then apply several SQL statements, and last issue a commit transaction command. Here we know all the action before actually executing them, so if there are multiple actions operation on the same row, we could merge them first. For example, if there are two increment on the same row, we could just add the values to merge them into one increment. And if there is a CAS and then a Put, we could just remove the CAS as the internal value is not visible right? And if there is a Put and then a CAS, it is easy to know whether the CAS can be performed, and can even convert it to a Put.

Or maybe even make a simple statement that, please do not operate on the same row in a batch, we will not consider the previous operation in the same batch when performing the operations in the batch.

Reply:
> Here we know all the action before actually executing them, so if there are multiple actions operation on the same row, we could merge them first.

I thought the same thing first, but it is not such a simple thing because we can specify timestamp to mutations. So we can't always know the latest column value without merging the existing values (in memstore and hfiles). For example, if we have one existing Put with timestamp 2, and if we perform Put with timestamp 1 and CAS for the same column in a batch, we can't know the latest column value without merging the existing values. That's why I needed to add the new additional Scanner. However, I know it will make things complicated.

> Or maybe even make a simple statement that, please do not operate on the same row in a batch, we will not consider the previous operation in the same batch when performing the operations in the batch.

Sounds this is a good idea. I will modify the patch this way. Thank you for discussing this.